### PR TITLE
rocjitsu: Harden ISA layer, virtual machine, and adds initial DBT

### PR DESCRIPTION
**Pull request by atgutier**
## Motivation

Provides many fixes and enhancements to establish our MVP:

Opt125 runs a forward pass on the simulator in the vllm-openai-rocm container for the cdna4 KMD config and produces the correct token:

Final logits: NaN=0 Inf=0 total=50272
GPU top-5: [(' the', '9.03'), (' is', '7.57'), (' a', '6.89'), (' has', '6.19'), (' "', '5.50')]

Loading model on CPU for comparison...
CPU top-5: [(' the', '7.66'), (' a', '7.50'), (' not', '6.82'), (' in', '6.61'), (' being', '6.24')]
Logit max_diff=9.1250 mean_diff=1.3923

DBT layer demonstrates capabilities to translate between CDNA4 native source binary and RDNA4 Wave64 mode target with correct results against a fuzzing test of random inputs.

## Technical Details

Many fixes to the ISA layer, pipeline, and KMD. Initial architecture in place for DBT.

## Test Plan

Run all tests in our suite and validate Opt125 run in the simulator on the vllm-openai-rocm container.

## Test Result

Tests all pass. Real GFX12 HW runs of translated binaries pass (requires GFX12 HW), Opt125 forward pass produces good token in Docker when run on CDNA4 KMD config.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

**Review by MaheshRavishankar — APPROVED** [review]
Could you please clean up the commit title...

This is experimental and reviewed through agents. I am fine with this landing and fixing forward any issues since this is MVP.

File: experimental/rocjitsu/docs/dbt-design.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+- Manage code caves for expanded instructions (branch stubs in .text, bodies in NOP padding)
+- Delegate ELF and kernel descriptor patching to `CodeObjectPatcher`
+
+### Key design constraint
+
+The binary translator's main loop is ISA-agnostic. It contains no conditional branches on `guest_arch_` or `host_arch_`, no ISA-specific register constants, no encoding format knowledge, and no instruction mnemonics. All ISA-specific logic lives in the encoding and semantic translators.
+
+### Code cave mechanism
+
+When a translated instruction sequence is larger than the source instruction, the binary translator creates a code cave: a branch stub replaces the original instruction slot, and the expanded sequence is placed in the NOP padding after `s_endpgm`. A return branch at the end of the cave jumps back to the next instruction. The `s_branch` target is computed as `(cave_offset - (branch_pc + 4)) / 4` per the AMDGPU branch encoding.
```
Comment (amd-arosa) [inline experimental/rocjitsu/docs/dbt-design.md:56]: do we know the nops are big enough

File: experimental/rocjitsu/tests/dbt/translate_test.cpp
Comment (amd-arosa) [inline experimental/rocjitsu/tests/dbt/translate_test.cpp:1]: what is the status of this? having a note somewhere would be nice.

**Review by amd-arosa — APPROVED** [review]

File: experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/patch/instruction_builder.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  case ROCJITSU_CODE_ARCH_RDNA4:
+    return 32;
+  default:
+    return 2;
+  }
+}
+
+/// @brief Get the s_nop opcode for a target ISA.
+[[nodiscard]] inline constexpr uint32_t sopp_op_nop([[maybe_unused]] rj_code_arch_t arch) {
+  return 0; // s_nop is opcode 0 on all ISAs
```
Comment (Groverkss) [inline experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/patch/instruction_builder.h:55]: I don't think this is true. From your code:

s_nop encodes as 0xBF800000 on both CDNA4 and RDNA4

File: experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/dbt/translation_rule.h [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+                                           const RegisterLiveness &liveness,
+                                           const LaneLayout *guest_layout,
+                                           const LaneLayout *host_layout);
+
+/// @brief A single translation rule for one (source, target) instruction.
+///
+/// @details Keyed by src_opcode for lookup via binary search. The expand_fn
+/// is called with the guest/host LaneLayout pointers for matrix instructions.
+struct TranslationRule {
+  uint16_t src_opcode; ///< Source opcode (primary key for lookup).
```
Comment (Groverkss) [inline experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/dbt/translation_rule.h]: I think two different instructions can have same opcodes but different encoding ids:

CDNA4 SOPP:S_WAITCNT and SOP2:S_AND_B32 both use opcode 12

so you need to use (encoding_id, op_code) as the translation key for lookup.

File: experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/dbt/generated/encoding_cdna4_to_rdna4.h [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+inline constexpr uint32_t kEnc_VOP3P = 0x1A7;
+inline constexpr uint32_t kEnc_VOP3_SDST_ENC = 0x1A2;
+
+inline TranslationResult translate_encoding_cdna4_to_rdna4(uint32_t encoding_id, uint32_t w0,
+                                                           uint32_t w1,
+                                                           [[maybe_unused]] uint32_t w2,
+                                                           uint16_t dst_op) {
+  switch (encoding_id) {
+  case kEnc_VOP2:
+    return encode_vop2_rdna4(decode_vop2_cdna4(w0), dst_op);
```
Comment (Groverkss) [inline experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/dbt/generated/encoding_cdna4_to_rdna4.h:545]: Side comment: I had to locally patch it to distinguish between sdwa/dpp encoding classes while decoding. Not sure what the best fix is however.

**Comment by Groverkss** [comment]
I left some comments on some potential bugs that I fixed locally, feel free to land without them and I can send a follow up patch for them.

File: experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/patch/code_object_patcher.cpp [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+      if (AMDHSA_BITS_GET(kcp, kd::KERNEL_CODE_PROPERTY_ENABLE_SGPR_QUEUE_PTR))
+        sgpr += 2;
+      if (AMDHSA_BITS_GET(kcp, kd::KERNEL_CODE_PROPERTY_ENABLE_SGPR_KERNARG_SEGMENT_PTR))
+        sgpr += 2;
+      if (AMDHSA_BITS_GET(kcp, kd::KERNEL_CODE_PROPERTY_ENABLE_SGPR_DISPATCH_ID))
+        sgpr += 2;
+      if (AMDHSA_BITS_GET(kcp, kd::KERNEL_CODE_PROPERTY_ENABLE_SGPR_FLAT_SCRATCH_INIT))
+        sgpr += 2;
+      if (AMDHSA_BITS_GET(kcp, kd::KERNEL_CODE_PROPERTY_ENABLE_SGPR_PRIVATE_SEGMENT_SIZE))
+        sgpr += 1;
```
Comment (Groverkss) [inline experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/patch/code_object_patcher.cpp]: You can use COMPUTE_PGM_RSRC2.USER_SGPR_COUNT here instead

Comment (Groverkss) [reply experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/patch/code_object_patcher.cpp]: From https://llvm.org/docs/AMDGPUUsage.html#initial-kernel-execution-state :

The initial SGPRs comprise up to 16 User SGPRs that are set by CP and apply to all wavefronts of the grid. It is possible to specify more than 16 User SGPRs using the enable_sgpr_* bit fields, in which case only the first 16 are actually initialized. These are then immediately followed by the System SGPRs that are set up by ADC/SPI and can have different values for each wavefront of the grid dispatch.

I had to do this to support preloaded kernel arguements.

File: experimental/rocjitsu/docs/dbt-design.md (hunk 1 above)
Comment (atgutier) [reply experimental/rocjitsu/docs/dbt-design.md:56]: No it's only for certain cases. Otherwise we can put the caves in our own section.

File: experimental/rocjitsu/lib/python/amdisa/codegen.py [hunk 6]
```diff
@@ -4402,32 +4888,18 @@ def _enc_has_semantics(self, enc: InstEncoding) -> bool:
         'vector_swap',
         # Vector readlane/writelane/readfirstlane access encoding fields:
         'vector_readlane', 'vector_writelane', 'vector_readfirstlane',
-        # Vector compares have different executions between RDNA and CDNA
+        # V_CMPX writes VCC+EXEC on CDNA but only EXEC on RDNA:
         'vector_cmpx', 'vector_cmpx_class',
     })
 
-    def _can_share_execute(self, mnemonic: str, encoding_name: str) -> bool:
+    def _can_share_execute(self, mnemonic: str) -> bool:
```
Comment (accauble) [inline experimental/rocjitsu/lib/python/amdisa/codegen.py:4895]: I think this function has been modified so that it will only return true if it is a universal mnemonic. Was that the intention? Or is a future PR going to undo that change?

If that was not the intention, here is the logic:

[PR 5129](https://github.com/ROCm/rocm-systems/pull/5129/) updated this to take both mnemonic and encoding_name as a key to be able to be able to share code between instructions with the same mnemonic but multiple encodings. The big bug was that the sharing was happening between the encodings when it shouldn't have been. I believe your code still prevents that from happening. However, it does remove the ability to family share with different encodings.

The shared_plan.family_shared still [is populated with (mnemonic, encoding)](https://github.com/ROCm/rocm-systems/blob/f3a88880c984e1772db29b86e1be99be4b969699/experimental/rocjitsu/lib/python/amdisa/cross_isa.py#L274) and so I believe [the check on line 4917](https://github.com/ROCm/rocm-systems/blob/f3a88880c984e1772db29b86e1be99be4b969699/experimental/rocjitsu/lib/python/amdisa/codegen.py#L4917) will always fail.

**Review by accauble — COMMENTED** [review]
Not done reviewing but since you're making changes, you may want to address this sooner rather than later.

File: experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/patch/instruction_builder.h (hunk 2 above)
Comment (atgutier) [reply experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/patch/instruction_builder.h:55]: Correct but its opcode is 0, so sopp_op_nop when packed up goes thru this:

/// @brief SOPP encoding prefix, consistent across all AMDGPU ISA generations.
inline constexpr uint32_t kSoppEncodingPrefix = 0x17F;

/// @brief Pack a SOPP instruction word from its constituent fields.
///
/// @param op      7-bit SOPP opcode.
/// @param simm16  16-bit immediate field.
/// @returns The encoded 32-bit instruction word.
[[nodiscard]] inline constexpr uint32_t pack_sopp(uint32_t op, uint16_t simm16) {
  return (kSoppEncodingPrefix << 23) | (op << 16) | simm16;
}

resolves to 0xBF800000

There may be a clearer way to do this.

File: experimental/rocjitsu/tests/dbt/translate_test.cpp
Comment (atgutier) [reply experimental/rocjitsu/tests/dbt/translate_test.cpp:1]: Added a comment.

File: experimental/rocjitsu/lib/python/amdisa/codegen.py (hunk 6 above)
Comment (atgutier) [reply experimental/rocjitsu/lib/python/amdisa/codegen.py:4895]: Fixed

File: experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/patch/code_object_patcher.cpp (hunk 5 above)
Comment (atgutier) [reply experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/patch/code_object_patcher.cpp]: Fixed

File: experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/dbt/translation_rule.h (hunk 3 above)
Comment (atgutier) [reply experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/dbt/translation_rule.h]: Fixed

File: experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/dbt/generated/encoding_cdna4_to_rdna4.h (hunk 4 above)
Comment (atgutier) [reply experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/code/dbt/generated/encoding_cdna4_to_rdna4.h:545]: There are has_dpp() and has_sdwa() that come straight from the MR ISA specs. We can expose them on the base Instruction object via our Flags.

File: experimental/rocjitsu/lib/python/amdisa/codegen.py [hunk 7]
```diff
@@ -3206,7 +3379,10 @@ def _gen_vector_cmpx(self, src: list[str], op: str | None, dtype: str | None, is
             L.append('      result |= (1ULL << lane);')
         L.append('  }')
         if self.isa_spec.profile.cmpx_writes_vcc:
-            L.append('  wf.set_vcc(result);')
+            if dst and is_vop3:
```
Comment (accauble) [inline experimental/rocjitsu/lib/python/amdisa/codegen.py:3382]: Super nit-picky, especially in the age of agents:

This line currently only happens if cmpx_writes_vcc is set, which is fine, but I'm worried that someone down the line could unset that bool if dst is_vop3 (since cmpx wouldn't actually be writing vcc). Could we add a test that makes sure the write_scalar64 gets written in that case?

File: experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/isa/arch/amdgpu/shared/execute_shared.h [hunk 8]
```diff
@@ -80,12 +142,14 @@ inline void execute_ds_swizzle_b32_ds([[maybe_unused]] Inst &inst, [[maybe_unuse
       continue;
     uint32_t src_lane;
     if (offset & 0x8000) {
+      // QDMode: swizzle within 4-lane quads.
```
Comment (accauble) [inline experimental/rocjitsu/lib/rocjitsu/src/rocjitsu/isa/arch/amdgpu/shared/execute_shared.h:145]: Looks like these got added back in :)

File: experimental/rocjitsu/lib/python/amdisa/codegen.py (hunk 6 above)
Comment (accauble) [inline experimental/rocjitsu/lib/python/amdisa/codegen.py:4895]: After investigating how this change affects the code generation, I am convinced that the encoding specifics are handled correctly after the function is called and this even opens up some functions for sharing that couldn't be shared before. I think because we can now share across rdna and cdna (e.g. `execute_v_dot2c_f32_f16_vop3` was shared by cdna1-3 and is now shared by rdna1 as well).

The change does break a test though: `test_v_cmpx_instructions_cannot_share_execute` in `test_cross_isa_analyzer.py` calls this function so you just nee to remove the encoding

**Review by newling — APPROVED** [review]
I haven't looked at this, but after rebasing onto this, issues I have on my race detection plugin development branch are resolved. So LGTM!
